### PR TITLE
[autoscaler] [azure] Fix Azure Autoscaling Failures

### LINKED ISSUE #16091: [autoscaler] No longer possible to create an Azure cluster
### What is the problem?

It is no longer possible to create an Azure cluster.

```
❯ ray up -y example-full.yaml
Cluster: default

Checking Azure environment settings
Traceback (most recent call last):
  File "/private/tmp/azure-test/venv/bin/ray", line 8, in <module>
    sys.exit(main())
  File "/private/tmp/azure-test/venv/lib/python3.8/site-packages/ray/scripts/scripts.py", line 1706, in main
    return cli()
  File "/private/tmp/azure-test/venv/lib/python3.8/site-packages/click/core.py", line 1137, in __call__
    return self.main(*args, **kwargs)
  File "/private/tmp/azure-test/venv/lib/python3.8/site-packages/click/core.py", line 1062, in main
    rv = self.invoke(ctx)
  File "/private/tmp/azure-test/venv/lib/python3.8/site-packages/click/core.py", line 1668, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/private/tmp/azure-test/venv/lib/python3.8/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/private/tmp/azure-test/venv/lib/python3.8/site-packages/click/core.py", line 763, in invoke
    return __callback(*args, **kwargs)
  File "/private/tmp/azure-test/venv/lib/python3.8/site-packages/ray/scripts/scripts.py", line 868, in up
    create_or_update_cluster(
  File "/private/tmp/azure-test/venv/lib/python3.8/site-packages/ray/autoscaler/_private/commands.py", line 239, in create_or_update_cluster
    config = _bootstrap_config(config, no_config_cache=no_config_cache)
  File "/private/tmp/azure-test/venv/lib/python3.8/site-packages/ray/autoscaler/_private/commands.py", line 315, in _bootstrap_config
    resolved_config = provider_cls.bootstrap_config(config)
  File "/private/tmp/azure-test/venv/lib/python3.8/site-packages/ray/autoscaler/_private/azure/node_provider.py", line 305, in bootstrap_config
    return bootstrap_azure(cluster_config)
  File "/private/tmp/azure-test/venv/lib/python3.8/site-packages/ray/autoscaler/_private/azure/config.py", line 21, in bootstrap_azure
    config = _configure_resource_group(config)
  File "/private/tmp/azure-test/venv/lib/python3.8/site-packages/ray/autoscaler/_private/azure/config.py", line 38, in _configure_resource_group
    subscription_id = resource_client.config.subscription_id
AttributeError: 'ResourceManagementClient' object has no attribute 'config'
```

*Ray version and other system information (Python version, TensorFlow version, OS):*

- OSX
- Python 3.8
- Ray 1.3.0 and 2.0.0.dev0
- Azure-cli 2.24.0

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

- Download the [Azure Cluster example file](https://raw.githubusercontent.com/ray-project/ray/master/python/ray/autoscaler/azure/example-full.yaml)
- Adjust `subscription_id`
- Install deps `pip install azure-cli-core azure-mgmt-compute azure-mgmt-msi azure-mgmt-network azure-mgmt-resource ray[default]`
- Execute `ray up -y example-full.yaml`

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by gramhagen**
- update node provider to support newer versions of azure-mgmt-resource sdk
 
azure autoscaler yaml updates: 
- use lastest vm image
- add pause after docker usermod for better robustness
- remove conda env in setup for docker based config
- update conda env to use python 3.8 versions

## Why are these changes needed?
ray up using azure yamls fails with various errors

## Related issue number

Closes #16091

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

I have manually tested using example_full.yaml, need to test other yamls (gpu, legacy, minimal)

### CODE REVIEW

**Comment by gramhagen** [comment]
seeing more issues when setting up worker nodes... do not merge this yet

**Comment by gramhagen** [comment]
I'm seeing a strange error on trying to deploy the worker nodes:
```
2021-06-24 01:19:02,577 INFO monitor.py:400 -- Starting monitor using ray installation: /anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/__init__.py
2021-06-24 01:19:02,577 INFO monitor.py:401 -- Ray version: 2.0.0.dev0
2021-06-24 01:19:02,577 INFO monitor.py:402 -- Ray commit: 3e2f6081459a4a6c79b33e633bffef1dc73be44b
2021-06-24 01:19:02,577 INFO monitor.py:403 -- Monitor started with command: ['/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/autoscaler/_private/monitor.py', '--logs-dir=/tmp/ray/session_2021-06-24_01-19-00_401345_10781/logs', '--redis-address=10.47.0.4:6379', '--logging-rotate-bytes=536870912', '--logging-rotate-backup-count=5', '--autoscaling-config=~/ray_bootstrap_config.yaml', '--redis-password=5241590000000000', '--monitor-ip=10.47.0.4']
2021-06-24 01:19:02,872 INFO monitor.py:143 -- Starting autoscaler metrics server on port 44217
2021-06-24 01:19:03,024 INFO monitor.py:151 -- Monitor: Started
2021-06-24 01:19:03,082 ERROR monitor.py:285 -- Error in monitor loop
Traceback (most recent call last):
  File "/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/autoscaler/_private/monitor.py", line 316, in run
    self._initialize_autoscaler()
  File "/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/autoscaler/_private/monitor.py", line 163, in _initialize_autoscaler
    prom_metrics=self.prom_metrics)
  File "/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/autoscaler/_private/autoscaler.py", line 95, in __init__
    self.reset(errors_fatal=True)
  File "/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/autoscaler/_private/autoscaler.py", line 573, in reset
    raise e
  File "/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/autoscaler/_private/autoscaler.py", line 525, in reset
    self.config["cluster_name"])
  File "/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/autoscaler/_private/providers.py", line 200, in _get_node_provider
    provider_cls = _get_node_provider_cls(provider_config)
  File "/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/autoscaler/_private/providers.py", line 176, in _get_node_provider_cls
    return importer(provider_config)
  File "/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/autoscaler/_private/providers.py", line 38, in _import_azure
    from ray.autoscaler._private.azure.node_provider import AzureNodeProvider
  File "/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/autoscaler/_private/azure/node_provider.py", line 7, in <module>
    from azure.common.client_factory import get_client_from_cli_profile
ModuleNotFoundError: No module named 'azure.common'

==> /tmp/ray/session_latest/logs/monitor.out <==
```

this doesn't  make much sense because if I attach to the head node I can start python and run `from azure.common.client_factory import get_client_from_cli_profile` without a problem. 
However if I try to execute `ray monitor ray_bootstrap_config.yaml` I get the same error. 

@AmeerHajAli do you know of any reason modules would not be accessible when running the ray package that is installed using this wheel: `pip install -U "ray[default] @ https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-2.0.0.dev0-cp37-cp37m-manylinux2014_x86_64.whl"`?

**Comment by AmeerHajAli** [comment]
@DmitriGekhtman @ijrsvt , can you please help @gramhagen here?

**Comment by gramhagen** [comment]
strangely, this doesn't occur when I use a previous wheel: https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-1.2.0.dev0-cp37-cp37m-manylinux2014_x86_64.whl

**Comment by philippeboyd** [comment]
@gramhagen I get the same error with Ray `1.4.0`

Head runs fine, but can't start a worker;  doing `from azure.common.client_factory import get_client_from_cli_profile`. directly on the head works like you said. I'm debugging at thee same time...

**Comment by ijrsvt** [comment]
@gramhagen Which YAML are you using to launch the cluster?

**Comment by philippeboyd** [comment]
@gramhagen it's essentially the same one as specified in the issue ([the azure example file](https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/azure/example-full.yaml)) except I don't install anything in the `setup_commands` since i already have everything in my docker image (ray `1.4.0`)

**Comment by gramhagen** [comment]
this yaml works, but swapping out to use latest 2.0.0 wheel fails
```
# An unique identifier for the head node and workers of this cluster.
cluster_name: default

# The maximum number of workers nodes to launch in addition to the head
# node.
max_workers: 2

# The autoscaler will scale up the cluster faster with higher upscaling speed.
# E.g., if the task requires adding more nodes then autoscaler will gradually
# scale up the cluster in chunks of upscaling_speed*currently_running_nodes.
# This number should be > 0.
upscaling_speed: 1.0

# This executes all commands on all nodes in the docker container,
# and opens all the necessary ports to support the Ray cluster.
# Empty object means disabled.
docker: {}

# If a node is idle for this many minutes, it will be removed.
idle_timeout_minutes: 5

# Cloud-provider specific configuration.
provider:
    type: azure
    # https://azure.microsoft.com/en-us/global-infrastructure/locations
    location: westus2
    resource_group: ray-cluster
    # set subscription id otherwise the default from az cli will be used
    # subscription_id: 00000000-0000-0000-0000-000000000000

# How Ray will authenticate with newly launched nodes.
auth:
    ssh_user: ubuntu
    # you must specify paths to matching private and public key pair files
    # use `ssh-keygen -t rsa -b 4096` to generate a new ssh key pair
    ssh_private_key: ~/.ssh/id_rsa
    # changes to this should match what is specified in file_mounts
    ssh_public_key: ~/.ssh/id_rsa.pub

# More specific customization to node configurations can be made using the ARM template azure-vm-template.json file
# See documentation here: https://docs.microsoft.com/en-us/azure/templates/microsoft.compute/2019-03-01/virtualmachines
# Changes to the local file will be used during deployment of the head node, however worker nodes deployment occurs
# on the head node, so changes to the template must be included in the wheel file used in setup_commands section below

# Tell the autoscaler the allowed node types and the resources they provide.
# The key is the name of the node type, which is just for debugging purposes.
# The node config specifies the launch config and physical instance type.
available_node_types:
    ray.head.default:
        # The minimum number of worker nodes of this type to launch.
        # This number should be >= 0.
        min_workers: 0
        # The maximum number of worker nodes of this type to launch.
        # This takes precedence over min_workers.
        max_workers: 0
        # The resources provided by this node type.
        resources: {"CPU": 2}
        # Provider-specific config, e.g. instance type.
        node_config:
            azure_arm_parameters:
                vmSize: Standard_D2s_v3
                # List images https://docs.microsoft.com/en-us/azure/virtual-machines/linux/cli-ps-findimage
                imagePublisher: microsoft-dsvm
                imageOffer: ubuntu-1804
                imageSku: 1804-gen2
                imageVersion: 21.01.21

    ray.worker.default:
        # The minimum number of nodes of this type to launch.
        # This number should be >= 0.
        min_workers: 1
        # The resources provided by this node type.
        resources: {"CPU": 2}
        # Provider-specific config, e.g. instance type.
        node_config:
            azure_arm_parameters:
                vmSize: Standard_D2s_v3
                # List images https://docs.microsoft.com/en-us/azure/virtual-machines/linux/cli-ps-findimage
                imagePublisher: microsoft-dsvm
                imageOffer: ubuntu-1804
                imageSku: 1804-gen2
                imageVersion: 21.01.21
                # optionally set priority to use Spot instances
                priority: Spot
                # set a maximum price for spot instances if desired
                # billingProfile:
                #     maxPrice: -1

# Specify the node type of the head node (as configured above).
head_node_type: ray.head.default

# Files or directories to copy to the head and worker nodes. The format is a
# dictionary from REMOTE_PATH: LOCAL_PATH, e.g.
file_mounts: {
#    "/path1/on/remote/machine": "/path1/on/local/machine",
#    "/path2/on/remote/machine": "/path2/on/local/machine",
     "~/.ssh/id_rsa.pub": "~/.ssh/id_rsa.pub"
}

# Files or directories to copy from the head node to the worker nodes. The format is a
# list of paths. The same path on the head node will be copied to the worker node.
# This behavior is a subset of the file_mounts behavior. In the vast majority of cases
# you should just use file_mounts. Only use this if you know what you're doing!
cluster_synced_files: []

# Whether changes to directories in file_mounts or cluster_synced_files in the head node
# should sync to the worker node continuously
file_mounts_sync_continuously: False

# Patterns for files to exclude when running rsync up or rsync down
rsync_exclude: []

# Pattern files to use for filtering out files when running rsync up or rsync down. The file is searched for
# in the source directory and recursively through all subdirectories. For example, if .gitignore is provided
# as a value, the behavior will match git's behavior for finding and using .gitignore files.
rsync_filter: []

# List of commands that will be run before `setup_commands`. If docker is
# enabled, these commands will run outside the container and before docker
# is setup.
initialization_commands:
    # get rid of annoying Ubuntu message
    - touch ~/.sudo_as_admin_successful

# List of shell commands to run to set up nodes.
setup_commands:
    # Note: if you're developing Ray, you probably want to create an AMI that
    # has your Ray repo pre-cloned. Then, you can replace the pip installs
    # below with a git checkout <your_sha> (and possibly a recompile).
    - (which conda && echo 'eval "$(conda shell.bash hook)"' >> ~/.bashrc) || true
    # - echo 'conda activate py37_pytorch' >> ~/.bashrc
    - (conda activate py37_tensorflow &> /dev/null && echo 'conda activate py37_tensorflow' >> ~/.bashrc) || true
    - which ray || pip install -U "ray[default] @ https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-1.4.0.dev0-cp37-cp37m-manylinux2014_x86_64.whl"
    # Consider uncommenting these if you also want to run apt-get commands during setup
    # - sudo pkill -9 apt-get || true
    # - sudo pkill -9 dpkg || true
    # - sudo dpkg --configure -a

# Custom commands that will be run on the head node after common setup.
head_setup_commands:
    - pip install -U azure-cli-core==2.22.0 azure-mgmt-compute==14.0.0 azure-mgmt-msi==1.0.0 azure-mgmt-network==10.2.0 azure-mgmt-resource==13.0.0

# Custom commands that will be run on worker nodes after common setup.
worker_setup_commands: []

# Command to start ray on the head node. You don't need to change this.
head_start_ray_commands:
    - ray stop
    - ulimit -n 65536; ray start --head --port=6379 --object-manager-port=8076 --autoscaling-config=~/ray_bootstrap_config.yaml

# Command to start ray on worker nodes. You don't need to change this.
worker_start_ray_commands:
    - ray stop
    - ulimit -n 65536; ray start --address=$RAY_HEAD_IP:6379 --object-manager-port=8076

head_node: {}
worker_nodes: {}
```

**Comment by ijrsvt** [comment]
Hmm @gramhagen I'm wondering if the autoscaler is starting without `conda` (i.e. that the conda environment is not being picked up by default).

**Comment by gramhagen** [comment]
if so it's happening inside the ray code somehow (maybe ray or click is kicking off a new thread or not using the same shell?)
i can reproduce the issue by using a docker based yaml which doesn't rely on conda. after attaching to the head node and confirming azure-common in installed and exists in the site-packages folder if I run
`ray monitor ray_bootstrap_config.yaml` the error occurs again.
whereas just running `python -c "from azure.common.client_factory import get_client_from_cli_profile"` there's no error.
super weird... I wouldn't think ray would somehow inject something into the pythonpath which would cause a collision with azure?

I actually tried modifying `/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/autoscaler/_private/azure/node_provider.py` on the head node to add prints before the import and after. What is strange to me is the file seems to be loaded twice, the first time there is no error (both prints come through), the second time only the first print happens and then the import fails.

**Comment by philippeboyd** [comment]
@gramhagen I went through the same kind of debugging process as you did. 

in the file `/usr/local/lib/python3.8/site-packages/ray/autoscaler/_private/monitor.py` I added the following import in first position just to test the import itself.
```
"""Autoscaler monitoring loop daemon."""
from azure.common.client_factory import get_client_from_cli_profile
# other imports...
```

and then simply callling `python /usr/local/lib/python3.8/site-packages/ray/autoscaler/_private/monitor.py`

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/ray/autoscaler/_private/monitor.py", line 2, in <module>
    from azure.common.client_factory import get_client_from_cli_profile
ModuleNotFoundError: No module named 'azure.common'
```

I get the same error. I then noticed that there's also a package named `azure` in `/usr/local/lib/python3.8/site-packages/ray/autoscaler/_private/`

I'm assuming there's probably a conflict with the relative package `azure` in `/usr/local/lib/python3.8/site-packages/ray/autoscaler/_private/`

When calling directly the file `python /usr/local/lib/python3.8/site-packages/ray/autoscaler/_private/azure/node_provider.py` I don't get the error...

**Comment by gramhagen** [comment]
ooof, yeah that makes sense. curious why this only started popping up recently though. i thought it has been like that for a while.  or maybe the bug has also been around for a while =/

seems like we could just rename the folder or do some hacky modification to the sys.path

**Comment by gramhagen** [comment]
i'm wondering now if that's why kubernetes is _kubernetes

**Comment by gramhagen** [comment]
updated pr here to rename folder

**Comment by ijrsvt** [comment]
@gramhagen Oh yes, that is why `kubernetes` was changed to `_kubernetes`. Python gets confused between the directory `azure` and the pip installed `azure` modules

**Comment by ijrsvt** [comment]
Thanks @gramhagen and @philippeboyd for figuring this out =)

**Comment by philippeboyd** [comment]
@gramhagen quick check on the modified files, I think you forgot azure's `node_provider.py` ?

```
from ray.autoscaler._private.azure.config import bootstrap_azure
```

**Comment by philippeboyd** [comment]
@gramhagen don't forget the `setup.py` file also. Disregard if already done
image

**Comment by gramhagen** [comment]
ack, thanks, i didn't think about search for actual folder paths

**Comment by richardliaw** [comment]
Is this ready to be merged?

**Comment by DmitriGekhtman** [comment]
There are some config test failures -- taking a look.

File: python/ray/autoscaler/azure/example-full-legacy.yaml
Comment (DmitriGekhtman) [inline python/ray/autoscaler/azure/example-full-legacy.yaml:132]: Need an empty list here, rather than nil.
`setup_commands: []`

**Comment by DmitriGekhtman** [comment]
Besides minor typo mentioned above, lgtm!
@ijrsvt could you take a final look?

**Review by ijrsvt — APPROVED** [review]
LGTM, pending the addition Dmitri mentioned!

File: python/ray/autoscaler/azure/example-full-legacy.yaml
Comment (gramhagen) [reply python/ray/autoscaler/azure/example-full-legacy.yaml:132]: thanks, updated!

**Comment by ijrsvt** [comment]
@gramhagen Just to double check, have you since tested:
[ ] gpu, 
[ ]legacy
[ ] minimal

**Comment by gramhagen** [comment]
I need to build a new docker image and wheel to properly test these scenarios. Is there a way to have the ci system build an wheel that can be used?

**Comment by DmitriGekhtman** [comment]
To test python changes, you can use setup commands to clone this branch and run `ray/python/ray/setup-dev.py` to replace the ray python libraries with the code in this branch, something like
```
setup_commands:
     - git clone --single-branch --branch fix_azure https://github.com/gramhagen/ray
     - pushd ray/python/ray && python setup-dev.py -y && popd
```
But, actually, given that there are changes to `setup.py` in this PR, I'm not sure if this would work. Definitely this shouldn't require rebuilding the Ray C++ source.
Any suggestions, @ijrsvt ?

**Comment by ijrsvt** [comment]
Hmm, yeah @DmitriGekhtman. I'm not 100% sure what the best way to test this is. I think it is worth giving that a shot @gramhagen  (the `setup-dev.py` method).

**Comment by gramhagen** [comment]
ok, testing now.
one thing that i noticed is the azure-mgmt-resource==13.0.0 package needs to be added to the docker ray-project/ray* docker images. should I add a separate pr for that?

**Comment by DmitriGekhtman** [comment]
@gramhagen please feel free to add the module to `ray/docker/ray-deps/Dockerfile` in this PR

**Comment by gramhagen** [comment]
@DmitriGekhtman, ok, I moved azure-mgmt-resource install to the Dockerfile (and removed it from the setup commands in the yamls). This does mean they won't work until the docker images are updated. Do you think that's safe or should I keep the install in the yaml too?

@ijrsvt I've tested the following yamls and they are working (using the setup-dev.py method and provided the package mentioned above is installed)

- [x] example-full-legacy
- [x] example-full
- [x] example-gpu-docker
- [ ] example-minimal

but when running example-minimal I get the following error
```
2021-07-06 22:52:56,796 ERROR services.py:1324 -- Failed to start the dashboard: Failed to start the dashboard, return code 0. The last 10 lines of /tmp/ray/session_2021-07-06_22-52-52_779842_16325/logs/dashboard.log:
  File "/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/new_dashboard/dashboard.py", line 215, in <module>
    log_dir=args.log_dir)
  File "/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/new_dashboard/dashboard.py", line 100, in __init__
    raise ex
  File "/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/new_dashboard/dashboard.py", line 91, in __init__
    build_dir = setup_static_dir()
  File "/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/new_dashboard/dashboard.py", line 51, in setup_static_dir
    "&& npm run build)", build_dir)
FrontendNotFoundError: [Errno 2] Dashboard build directory not found. If installing from source, please follow the additional steps required to build the dashboard(cd python/ray/new_dashboard/client && npm install && npm ci && npm run build): '/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/new_dashboard/client/build'
[2021-07-06 22:53:01,840 C 16325 16325] service_based_gcs_client.cc:72:  Check failed: GetGcsServerAddressFromRedis( redis_client_->GetPrimaryContext()->sync_context(), &current_gcs_server_address_, RayConfig::instance().gcs_service_connect_retries()) Failed to get gcs server address when init gcs client.
```

I've tried running the npm commands to build the dashboard, but that doesn't fix the gcs client timeout
any thoughts on how to resolve this?

**Comment by richardliaw** [comment]
hmm, is it possible that there's a missing dependency?

or that it's only a transient error (does it happen multiple times)?

**Comment by DmitriGekhtman** [comment]
> Do you think that's safe or should I keep the install in the yaml too?

Should be safe to have it just in the Dockerfile, except for defaults.yaml which doesn't use Docker -- for defaults, it needs to be in the yaml.

> any thoughts on how to resolve this?

There was some weirdness around the dashboard in Ray master yesterday. This might be resolved now?
example-minimal.yaml pulls defaults from defaults.yaml, which includes a setup command to install Ray master.
defaults.yaml (and thus example-minimal.yaml) are weird in that they don't use docker.

If you're still seeing the issue with example-minimal, could you post how you're modifying configs to test example-minimal?

**Comment by gramhagen** [comment]
> > Do you think that's safe or should I keep the install in the yaml too?
> 
> Should be safe to have it just in the Dockerfile, except for defaults.yaml which doesn't use Docker -- for defaults, it needs to be in the yaml.

yeah, I kept the installation in defaults.yaml

> There was some weirdness around the dashboard in Ray master yesterday. This might be resolved now?
> example-minimal.yaml pulls defaults from defaults.yaml, which includes a setup command to install Ray master wheels.
> defaults.yaml (and thus example-minimal.yaml) are weird in that they don't use docker.
> 
> If you're still seeing the issue with example-minimal, could you post how you're modifying configs to test example-minimal?

that was it, I just merged in the latest changes and this went away, minimal now works too. so @richardliaw I think we're good with this one. Thanks for the help everyone!

File: docker/ray-deps/Dockerfile
Comment (ijrsvt) [inline docker/ray-deps/Dockerfile:20]:
```suggestion
        "azure-mgmt-network==10.2.0" \
```

File: python/ray/autoscaler/azure/example-full-legacy.yaml
Comment (ijrsvt) [inline python/ray/autoscaler/azure/example-full-legacy.yaml:139]:
```suggestion
head_setup_commands: []
```

Just to check, this is intentionally commented out still?

File: docker/ray-deps/Dockerfile
Comment (gramhagen) [reply docker/ray-deps/Dockerfile]: got it

File: python/ray/autoscaler/azure/example-full-legacy.yaml
Comment (gramhagen) [reply python/ray/autoscaler/azure/example-full-legacy.yaml]: arg no, fixed all these now

**Comment by ijrsvt** [comment]
Test failures are unrelated (`test_serve_deploy.py`).

@DmitriGekhtman are you okay if I merge?

**Comment by DmitriGekhtman** [comment]
> Test failures are unrelated (`test_serve_deploy.py`).
> 
> @DmitriGekhtman are you okay if I merge?

Looks good, yes!

**Pull request merged**
